src/main/java/com/redhat/jenkins/plugins/cachet/pipeline/CachetGatingWaitForResourcesStep.java

```diff
@@ -0,0 +1,163 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+public class CachetGatingWaitForResourcesStep extends Step implements Serializable {
+
+    private final List<String> resources;
+
+    @DataBoundConstructor
+    public CachetGatingWaitForResourcesStep(List<String> resources) {
+        this.resources = new ArrayList<>(resources);
+    }
+
+    public List<String> getResources(){
+        return resources;
+    }
+
+    @Override
+    public StepExecution start(StepContext context) throws Exception {
+        return new Execution(resources, context);
+    }
+
+    public static class Execution extends SynchronousStepExecution<Map<String, Long>> {
+
+        private transient final List<String> resources;
+
+        private static final float RECURRENCE_PERIOD_BACKOFF = 1.2f;
+        static final long MIN_RECURRENCE_PERIOD = 250; // ¼s
+        static final long MAX_RECURRENCE_PERIOD = 15000; // ¼min
+        long recurrencePeriod = MIN_RECURRENCE_PERIOD;
+
+        Execution(List<String> resources, StepContext context) {
+            super(context);
+            this.resources = resources;
+        }
+
+        @Override
+        protected Map<String, Long> run() throws Exception {
+            Map<String, Long> waitForResourcesMet = new HashMap<>();
+            List<String> resourcesToWaitFor = new ArrayList<>();
+
+            Map<String, CachetGatingMetrics> metrics = getCurrentMetrics();
+
+            for (String resource : resources){
+                if(!metrics.containsKey(resource))
+                    throw new AbortException("One or more resources you want to wait for does not exist.");
+
+                CachetGatingMetrics tmpResourceMet = metrics.get(resource);
+                if(tmpResourceMet.getGatingStatus().equals("Operational")){
+                    waitForResourcesMet.put(resource, tmpResourceMet.getGatedTimeElapsed());
+                }
+                else {
+                    resourcesToWaitFor.add(resource);
+                }
+            }
+            
+            while (!resourcesToWaitFor.isEmpty()){// So there are resources we need to wait for
+                sleepingHandler(resourcesToWaitFor);
+                for (String resource : resourcesToWaitFor){
+                    CachetGatingMetrics tmpResourceMet = getCurrentMetrics().get(resource);
+                    if(tmpResourceMet.getGatingStatus().equals("Operational")){
+                        resourcesToWaitFor.remove(resource);
```

Comment:
I am convinced removing from collection you are iterating over this way will blow up: https://stackoverflow.com/questions/223918/iterating-through-a-collection-avoiding-concurrentmodificationexception-when-re

Reply:
Thanks @olivergondza, I'll take care of that!